File: Source/Frontend/services/catalog/service-types-api.ts
import axiosClient from "@/lib/axios-interceptor";

export interface ServiceTypeFilterParams {
  page?: number;
  pageSize?: number;
  keyword?: string | null;
  langId?: number | null;
}

export interface CreateServiceTypeRequest {
  serviceTypeName: string;
  langId?: number | null;
}

export interface UpdateServiceTypeRequest {
  serviceTypeName: string;
  langId?: number | null;
}

export interface ServiceTypeResponse {
  serviceTypeId: number;
  serviceTypeName?: string | null;
  langId?: number | null;
  serviceCount: number;
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

const BASE_URL = "/catalog/service-types";

export const serviceTypeApi = {
  getList: (params: ServiceTypeFilterParams) => {
    return axiosClient.get<any, PagedResult<ServiceTypeResponse>>(BASE_URL, {
      params,
    });
  },
  getById: (id: number) => {
    return axiosClient.get<any, ServiceTypeResponse>(`${BASE_URL}/${id}`);
  },
  create: (data: CreateServiceTypeRequest) => {
    return axiosClient.post<any, ServiceTypeResponse>(BASE_URL, data);
  },
  update: (id: number, data: UpdateServiceTypeRequest) => {
    return axiosClient.put<any, { serviceTypeId: number }>(`${BASE_URL}/${id}`, data);
  },
  delete: (id: number) => {
    return axiosClient.delete<any, { serviceTypeId: number }>(`${BASE_URL}/${id}`);
  },
};

File: Source/flutter/lib/services/catalog/service_types_api.dart
import '../../models/catalog/service_type_dto.dart';
import '../api_client.dart';

class ServiceTypeApi {
  static final ApiClient _apiClient = ApiClient();
  static const String _baseUrl = '/catalog/service-types';

  /// Lấy danh sách loại dịch vụ có phân trang & tìm kiếm
  /// Mirrors: serviceTypeApi.getList(params: ServiceTypeFilterParams)
  static Future<PagedResult<ServiceTypeResponse>> getList(
    ServiceTypeFilterParams params,
  ) async {
    final queryParams = params.toQueryParameters();
    final queryString = Uri(queryParameters: queryParams).query;
    final endpoint = queryString.isNotEmpty ? '$_baseUrl?$queryString' : _baseUrl;

    final response = await _apiClient.get(endpoint);
    return PagedResult<ServiceTypeResponse>.fromJson(
      response as Map<String, dynamic>,
      (json) => ServiceTypeResponse.fromJson(json),
    );
  }

  /// Lấy chi tiết loại dịch vụ theo ID
  /// Mirrors: serviceTypeApi.getById(id: number)
  static Future<ServiceTypeResponse> getById(int id) async {
    final response = await _apiClient.get('$_baseUrl/$id');
    return ServiceTypeResponse.fromJson(response as Map<String, dynamic>);
  }

  /// Tạo loại dịch vụ mới
  /// Mirrors: serviceTypeApi.create(data: CreateServiceTypeRequest)
  static Future<ServiceTypeResponse> create(
    CreateServiceTypeRequest data,
  ) async {
    final response = await _apiClient.post(_baseUrl, body: data.toJson());
    return ServiceTypeResponse.fromJson(response as Map<String, dynamic>);
  }

  /// Cập nhật loại dịch vụ
  /// Mirrors: serviceTypeApi.update(id: number, data: UpdateServiceTypeRequest)
  static Future<void> update(
    int id,
    UpdateServiceTypeRequest data,
  ) async {
    await _apiClient.put('$_baseUrl/$id', body: data.toJson());
  }

  /// Xóa loại dịch vụ theo ID
  /// Mirrors: serviceTypeApi.delete(id: number)
  static Future<void> delete(int id) async {
    await _apiClient.delete('$_baseUrl/$id');
  }

  /// Xóa nhiều loại dịch vụ hàng loạt
  static Future<void> deleteBulk(List<int> ids) async {
    await Future.wait(ids.map((id) => delete(id)));
  }
}
